packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -24,71 +25,154 @@ enum TextSelectionHandleType { left, right, collapsed }
 /// Builds a handle of the given type.
 typedef Widget TextSelectionHandleBuilder(BuildContext context, TextSelectionHandleType type);
 
+// Builds a copy/paste toolbar.
+// TODO(mpcomplete): A single position is probably insufficient.
+typedef Widget TextSelectionToolbarBuilder(BuildContext context, Point position, TextSelectionDelegate delegate);
+
 /// The text position that a give selection handle manipulates. Dragging the
 /// [start] handle always moves the [start]/[baseOffset] of the selection.
 enum _TextSelectionHandlePosition { start, end }
 
+/// An interface for manipulating the selection, to be used by the implementor
+/// of the toolbar widget.
+abstract class TextSelectionDelegate {
+  /// Gets the current text input.
+  InputValue get inputValue;
+
+  /// Sets the current text input (replaces the whole line).
+  void set inputValue(InputValue value);
+
+  /// The copy/paste buffer. Application-wide.
+  String get pasteBuffer;
```

Comment:
This should return a `Future<String>` because  reading the platform clipboard might be asynchronous.